Compiler segfault while building roclang builtins (regression from zig 0.9)

### Zig Version

0.10.1

### Steps to Reproduce and Observed Behavior

1. Fetch the sources from https://github.com/roc-lang/roc/tree/main/crates/compiler/builtins/bitcode
2. zig build-obj src/main.zig -lc -fstrip -ffunction-sections -OReleaseFast
3. observe segfault (expect to see some sort of error)
4. upgrade to nightly zig 0.11.0-dev.3664+0f5aff344
5. Observe a new, different cryptic error - apparently now a for loop can not capture index?
```
 src/main.zig:85:30: error: extra capture in for loop
 inline for (INTEGERS) |T,i| {
```
on the bright side, this is better than segfault.


### Expected Behavior

1. I'd expect the compiler to not segfault
2. The spec says nothing about removal of the index capture in loops (if it does, sorry, I must have missed it).

On zig version 0.9 the build completes normally, as expected. Naturally, there are incompatibilities between compiler versions, but segfaults is probably not "to spec" either way.

--- a/src/Module.zig
+++ b/src/Module.zig
@@ -6636,12 +6636,14 @@ fn markDeclIndexAlive(mod: *Module, decl_index: Decl.Index) Allocator.Error!void
 }
 
 pub fn addGlobalAssembly(mod: *Module, decl_index: Decl.Index, source: []const u8) !void {
-    try mod.global_assembly.ensureUnusedCapacity(mod.gpa, 1);
-
-    const duped_source = try mod.gpa.dupe(u8, source);
-    errdefer mod.gpa.free(duped_source);
-
-    mod.global_assembly.putAssumeCapacityNoClobber(decl_index, duped_source);
+    const gop = try mod.global_assembly.getOrPut(mod.gpa, decl_index);
+    if (gop.found_existing) {
+        const new_value = try std.fmt.allocPrint(mod.gpa, "{s}\n{s}", .{ gop.value_ptr.*, source });
+        mod.gpa.free(gop.value_ptr.*);
+        gop.value_ptr.* = new_value;
+    } else {
+        gop.value_ptr.* = try mod.gpa.dupe(u8, source);
+    }
 }
 
 pub fn wantDllExports(mod: Module) bool {

--- a/test/behavior/asm.zig
+++ b/test/behavior/asm.zig
@@ -12,6 +12,9 @@ comptime {
     {
         asm (
             \\.globl this_is_my_alias;
+        );
+        // test multiple asm per comptime block
+        asm (
             \\.type this_is_my_alias, @function;
             \\.set this_is_my_alias, derp;
         );

